### PR TITLE
MAINT/ENH: stats.multivariate_normal.rvs: fix shape and speed up with Covariance

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
gh-16278
https://github.com/scipy/scipy/pull/12312#issuecomment-640298227 (direct call to BLAS can come later...)
supersedes gh-12184 

#### What does this implement/fix?
`stats.multivariate_normal.rvs` generates random variates using the Generator's `multivariate_normal`, which doesn't cache its factorization of the covariance matrix, so repeated calls require refactorizing the covariance matrix whether the distribution is frozen or not. The `Covariance` objects remember their factorization whether the distribution is frozen or not. 

Also, `stats.multivariate_normal.rvs` does an "indiscriminate squeeze", so singleton dimensions requested by the user are not present in the output. When the user passes an array as the covariance matrix, it is not easy to fix this in a backward-compatible way. But if the user passes a new `Covariance` instance, it might as well return an array of the correct shape.

This PR adds a `colorize` method to the `Covariance` class and uses it to generate random variates in `multivariate_normal.rvs`.

```python3
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt

rng = np.random.default_rng(984261616848)
size = (10, 10, 10)

mean = [10, 20]
V = np.array([[1, 1], [-1, 1]]) * np.sqrt(2)/2
W = np.diag([1, 5])
A = V.T @ W @ V

rvs1 = stats.multivariate_normal.rvs(mean, A, size=size, random_state=rng)
rvs1 = rvs1.reshape(1000, -1)

cov = stats.Covariance.from_cholesky(np.linalg.cholesky(A))
# cov = stats.Covariance.from_precision(np.linalg.inv(A))
# cov = stats.Covariance.from_eigendecomposition(np.linalg.eigh(A))
rvs2 = stats.multivariate_normal.rvs(mean, cov, size=size, random_state=rng)
rvs2 = rvs2.reshape(1000, -1)

plt.plot(*rvs1.T, '.', *rvs2.T, '.', alpha=0.5)
# plt.plot(*rvs2.T, '.', *rvs1.T, '.', alpha=0.5)  # looks the same
```

image

<details>

Even in low dimensional spaces and considering the cost of `Covariance` construction, the speed of the `Covariance` versions is comparable to that of NumPy's generator, which we currently rely on.

```python3
import numpy as np
from scipy import stats
rng = np.random.default_rng(984261616848)

mean = rng.random(10)
A = np.random.rand(10, 10)
A = A @ A.T

# One sample
%timeit rng.multivariate_normal(mean, A, size=None)
87 µs ± 811 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%timeit cov = stats.Covariance.from_cholesky(np.linalg.cholesky(A)); stats.multivariate_normal.rvs(mean, cov, size=None, random_state=rng)
# 45.8 µs ± 107 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%timeit cov = stats.Covariance.from_eigendecomposition(np.linalg.eigh(A)); stats.multivariate_normal.rvs(mean, cov, size=None, random_state=rng)
# 116 µs ± 1.2 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

# Many samples
size = (10, 10, 10)

%timeit rng.multivariate_normal(mean, A, size=size)
# 207 µs ± 7.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%timeit cov = stats.Covariance.from_cholesky(np.linalg.cholesky(A)); stats.multivariate_normal.rvs(mean, cov, size=size, random_state=rng)
# 205 µs ± 734 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%timeit cov = stats.Covariance.from_eigendecomposition(np.linalg.eigh(A)); stats.multivariate_normal.rvs(mean, cov, size=size, random_state=rng)
# 292 µs ± 419 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

The situation is even better in high-dimensional spaces.
```python3
n = 1000
mean = rng.random(n)
A = np.random.rand(n, n)
A = A @ A.T

%timeit rng.multivariate_normal(mean, A, size=None)
# 165 ms ± 503 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

%timeit cov = stats.Covariance.from_cholesky(np.linalg.cholesky(A)); stats.multivariate_normal.rvs(mean, cov, size=None, random_state=rng)
# 16.3 ms ± 92.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit cov = stats.Covariance.from_eigendecomposition(np.linalg.eigh(A)); stats.multivariate_normal.rvs(mean, cov, size=None, random_state=rng)
# 59.5 ms ± 1.81 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

And of course, if the covariance matrix is diagonal, the `Covariance` object created by `from_diagonal` is more accurate and faster still.

</details>

#### Additional information
Should the method be called `color` or `colorize`?

We'll fix the shape of `pdf`/`logpdf` output in a separate PR. We can use `Covariance` similarly in a few other multivariate distributions. IIUC, part of the speed increase is because these never use SVD. If we need a `from_SVD` version, it can easily be added.

@siddhantwahal

### CODE REVIEW

File: scipy/stats/_covariance.py
Comment (mdhaber) [inline scipy/stats/_covariance.py:388]: Elimination of the `dimensionality` parameter is a separate thing. We can move it to another PR, but if it's easy enough to include here, note that it's all contained in the first commit, so it can be filtered out by reviewing the commits after the first separately.

File: scipy/stats/tests/test_multivariate.py
Comment (tirthasheshpatel) [inline scipy/stats/tests/test_multivariate.py:177]:
```suggestion
            assert_equal(x1.shape, x.shape)
            assert_equal(x2.shape, x.shape)
```

Better to use `assert_equal` since they are all integers and should exactly match.

File: scipy/stats/tests/test_multivariate.py
Comment (tirthasheshpatel) [inline scipy/stats/tests/test_multivariate.py:124]:
```suggestion
        assert_equal(cov_object.rank, psd.rank)
```

File: scipy/stats/_covariance.py
Comment (tirthasheshpatel) [inline scipy/stats/_covariance.py:386]: Can you please also add https://www.projectrhea.org/rhea/index.php/ECE662_Whitening_and_Coloring_Transforms_S14_MH? It is cited in Wikipedia and is much more helpful.

File: scipy/stats/_covariance.py
Comment (tirthasheshpatel) [inline scipy/stats/_covariance.py:485]: Would this work for stacked inputs of shape `(x1, x2, ..., n, n)`? Or do you want to add support for that later?

Comment (mdhaber) [reply scipy/stats/_covariance.py:485]: Good catch. We need to add it here, and it's easy to do. The tests were supposed to test that... I'll have to take a closer look.

Comment (mdhaber) [reply scipy/stats/_covariance.py:485]: I confused myself. It does work and the test is working. I edited the code at the top to demonstrate. Please do try it!

File: scipy/stats/_covariance.py
Comment (mdhaber) [reply scipy/stats/_covariance.py:386]: I'm a little hesitant because *looks* a little less reputable than Wikipedia (which is a low bar.. even though Wikipedia cites it), and I didn't use it. "Coloring transform" is defined in the Wikipedia article, and based on that it must be the inverse of the whitening transform, so I thought that was enough. But if you ask again I'll add it : )

**Comment by mdhaber** [comment]
Thanks @tirthasheshpatel. I responded to the comments; I'll wait until the review is complete before pushing changes since they are small.

**Review by siddhantwahal — COMMENTED** [review]
Thanks for this work @mdhaber. The approach looks sound, I only have a few clarifying questions:

File: scipy/stats/_covariance.py
Comment (tirthasheshpatel) [reply scipy/stats/_covariance.py:386]: It might be useful for someone unaware of the concept since Wikipedia just writes one line about it and links to the article anyways. Also, I wasn't able to find anything else that talks about it in detail. If you can, feel free to add it instead!

File: scipy/stats/_multivariate.py
Comment (mdhaber) [reply scipy/stats/_multivariate.py:753]: > to start returning the correct shape to users when they specify a Covariance object explicitly.

Yes

> fix the incorrect shape for CovViaPSD at a later date.

Well, fix the incorrect shape for when the user passes in a regular array at a later day, yes. `CovViaPSD` is not public; it is only used when a user passes in a regular array instead of a `Covariance` object.

> Alternatively, for now, we could continue returning the incorrect shape for all covariance types and fix them (with due warning) later.

Backward incompatible changes are very not fun to work on. They are far more work than actually fixing the code, and then you have to wait a year to see real change.

So my plan now, personally, is to never fix the old classes and just introduce new ones, since there are several other problems (gh-16278). 

> That would keep the behavior across covariance types consistent at all times and avoid the need for branching (which will also be needed in other methods like pdf, logpdf?)

The only branching that needs to happen in the other methods is like:

```python3
if isinstance(cov_object, _covariance.CovViaPSD):
    out = _squeeze_output(out)
```

I'd still branch in this method because the intent was to keep using the same NumPy generator if the user passes in the same regular array. We don't *have* to maintain the same RVS stream for the same seed between feature releases, but I figured we might as well.

> Apologies if this approach was considered and rejected, I couldn't find it in the linked issues.

Yes, well, I considered it. Leaving the squeeze in there for everything is the default, and I went through a little extra trouble to fix the bug when users opt into using the `Covariance` class.

File: scipy/stats/_covariance.py
Comment (tirthasheshpatel) [inline scipy/stats/_covariance.py:365]:
```suggestion
        "Colorizing" ("color" as in "colored noise", in which different
```

**Review by tirthasheshpatel — COMMENTED** [review]
OK, I think this looks good other than the comments above.

File: scipy/stats/_multivariate.py
Comment (siddhantwahal) [reply scipy/stats/_multivariate.py:753]: That sounds good, thanks for the clarification.

File: scipy/stats/tests/test_multivariate.py
Comment (mdhaber) [reply scipy/stats/tests/test_multivariate.py:177]: Done in next commit.

File: scipy/stats/tests/test_multivariate.py
Comment (mdhaber) [reply scipy/stats/tests/test_multivariate.py:124]: Done in next commit.

File: scipy/stats/_covariance.py
Comment (mdhaber) [reply scipy/stats/_covariance.py:386]: Added in next commit: https://www.researchgate.net/publication/331063251_Generalization_of_Coloring_Linear_Transformation/link/5c63d0f345851582c3e42610/download

File: scipy/stats/_covariance.py
Comment (mdhaber) [inline scipy/stats/_covariance.py:340]:
```suggestion
        .. [2] Novak, Lukas, and Miroslav Vorechovsky. "Generalization of
```

File: scipy/stats/_covariance.py
Comment (mdhaber) [inline scipy/stats/_covariance.py:393]:
```suggestion
        .. [2] Novak, Lukas, and Miroslav Vorechovsky. "Generalization of
```

**Review by mdhaber — COMMENTED** [review]
These names appear without accents in the paper, so presumably it doesn't bother the authors much. This should fix CI, so if this is the last thing to fix, we can probably commit with `[skip ci]` and merge. Alternatively, if CI won't complain about this lint issue after it appears in `main`, we could just merge as-is.

**Comment by tirthasheshpatel** [comment]
> Alternatively, if CI won't complain about this lint issue after it appears in `main`, we could just merge as-is.

That sounds good. Main doesn't run the lint check so should not be a problem.

File: scipy/stats/_covariance.py
Comment (tirthasheshpatel) [reply scipy/stats/_covariance.py:389]: Thanks for lookin into this. The reference looks good!

**Review by tirthasheshpatel — APPROVED** [review]
I think this is ready to go in! Thanks @mdhaber @siddhantwahal!

**Comment by mdhaber** [comment]
Thanks @siddhantwahal @tirthasheshpatel!

**Pull request merged**
